pkg/os/util.go

```diff
@@ -92,9 +92,9 @@ func WriteFileIfContentChanged(path string, newContent []byte, perm os.FileMode)
 }
 
 func FileExists(path string) bool {
-	_, err := os.Stat(path)
-	if err != nil && os.IsNotExist(err) {
+	info, err := os.Lstat(path)
+	if err != nil {
 		return false
 	}
-	return true
+	return !info.IsDir()
```

Comment:
@cfergeau Even this change doesn't provide the correct behavior. (Trying with something which a user I don't have access to the folder and existing implementation did provide correct result.)

```
$ ls /var/lib/libvirt/qemu/
ls: cannot open directory '/var/lib/libvirt/qemu/': Permission denied

$ sudo ls /var/lib/libvirt/qemu/
channel  checkpoint  domain-2-crc-l44pc-master-0  domain-48-test1-44tjj-master-0  dump	nvram  ram  save  snapshot

$ cat fileexist.go
package main

import (
	"fmt"
	"os"
)

func FileExistsOld(path string) bool {
	_, err := os.Stat(path)
	if err != nil && os.IsNotExist(err) {
		return false
	}
	return true
}

func FileExistsNew(path string) bool {
	info, err := os.Lstat(path)
	if err != nil {
		return false
	}
	return !info.IsDir()
}

func main() {
	fmt.Println("New change in FileExist: ", FileExistsNew("/var/lib/libvirt/qemu/domain-48-test1-44tjj-master-0"))
	fmt.Println("Existing in FileExist: ", FileExistsOld("/var/lib/libvirt/qemu/domain-48-test1-44tjj-master-0"))
}

$ go run exist.go 
New change in FileExist:  false
Existing in FileExist:  true
```

Reply:
Since you have this:
```
$ ls /var/lib/libvirt/qemu/
ls: cannot open directory '/var/lib/libvirt/qemu/': Permission denied
```
the correct answer to `FileExists("/var/lib/libvirt/qemu/...")` is "I don't know" (in other words, there was an error, so I can't tell if the file exists or not). The old implementation will always return `true` when there is an error, regardless of the actual file existence. In your test case, `FileExistsOld("/var/lib/libvirt/qemu/domain-48-test1-44tjj-master-0")` returns true, but `FileExistsOld("/var/lib/libvirt/qemu/I know there is no such file")` woul also be true. The new implementation will always return `false" in such cases, which I believe is more correct given how the rest of the code uses this method.

Old implementation can have false positives (returns true when the file does not exist), new implementation can have false negatives (returns false when the file actually exists).
The correct implementation would be `FileExists(string) (bool, error)`, but this makes it less convenient to use:
```
exists, err := FileExists(path)
if err != nil {
   // do something which can be exists = false
}
if exists {
  // do something
}
``
as opposed to
```
if FileExists(path) {
  // do something
}
```

I preferred to keep the convenient API here.

Reply:
@cfergeau Can you provide a sample example with this false positive, I tired it with file which I don't have permission to list and vice versa, I didn't get any false positive, may be I am missing some test case here?

Reply:
In your previous test case, I'd expect this to return `true` even if `this is a random file` does not exist:
```
fmt.Println("Existing in FileExist: ", FileExistsOld("/var/lib/libvirt/qemu/this is a random file"))
```

Actually, I put a test case for this:

```
	err = os.Chmod(dirname, 0000)
	assert.NoError(t, err)
	assert.False(t, FileExists(filename))
	filename = filepath.Join(dirname, "nonexistent")
	assert.False(t, FileExists(filename))
```

you can try `go test ./pkg/os/` with the new/old implementation, and see the differences. With the old implementation, the 2 `assert.False(t, FileExists(...))` above will fail (ie `FileExists` returns `true` in these cases)